### PR TITLE
dialects (arm): add func op

### PR DESCRIPTION
**Pull request by emmau678**
Add funcOp for ARM, similar to that already in RISCV.

### CODE REVIEW

**Comment by emmau678** [comment]
thanks, could you please check if I've done that correctly when you have a chance? I just added an extra assertion to the existing pytest

File: xdsl/dialects/arm_func.py [hunk 1]
```diff
@@ -40,12 +151,13 @@ def assembly_line_args(self):
         return ()
 
     def assembly_instruction_name(self) -> str:
-        return "bx lr"
+        return "ret"
```
Comment (alexarice) [inline xdsl/dialects/arm_func.py:154]: These feel like a separate change?

File: xdsl/dialects/arm_func.py [hunk 2]
```diff
@@ -1,26 +1,137 @@
...
+            raise NotImplementedError("arg_attrs not implemented in riscv_func")
+        func = FuncOp(name, region, (input_types, return_types), visibility)
+        if extra_attrs is not None:
+            func.attributes |= extra_attrs.data
+        return func
+
+    def print(self, printer: Printer):
+        if self.sym_visibility:
+            visibility = self.sym_visibility.data
+            printer.print(f" {visibility}")
```
Comment (alexarice) [inline xdsl/dialects/arm_func.py:106]:
```suggestion
            printer.print_string(f" {visibility}")
```

**Review by alexarice — COMMENTED** [review]
Is there a good reason for allowing a region with no blocks? As far as I can tell it seems this has no representation in the assembly.

File: xdsl/dialects/arm_func.py (hunk 1 above)
Comment (alexarice) [reply xdsl/dialects/arm_func.py:154]: "these" being this and the comment above.

**Comment by superlopuh** [comment]
A function operation with a region with no blocks is a function declaration.

**Comment by alexarice** [comment]
So it's necessary in mlir just to declare the symbol?

**Comment by superlopuh** [comment]
I'm not 100% sure about this, but I think MLIR verifies that you only call functions that are declared in the module, so it's a bit of a consistency check. Also, I think that the external functions may have an assembly representation, for use during linking, but these also seem optional.

**Review by alexarice — APPROVED** [review]

**Pull request merged**